lib/matplotlib/collections.py

```diff
@@ -1105,76 +1105,105 @@ def legend_elements(self, prop="colors", num="auto",
-            label_values = loc.tick_values(func(arr).min(), func(arr).max())
-            cond = ((label_values >= func(arr).min()) &
-                    (label_values <= func(arr).max()))
+
+        if label_values_are_numeric:
+            label_values_min = label_values.min()
+            label_values_max = label_values.max()
+            fmt.set_bounds(label_values_min, label_values_max)
+
+            if num is not None:
+                # Labels are numerical but larger than the target
+                # number of elements, reduce to target using interpolation:
+                if isinstance(num, mpl.ticker.Locator):
+                    loc = num
+                elif np.iterable(num):
+                    loc = mpl.ticker.FixedLocator(num)
+                else:
+                    num = int(num)
+                    loc = mpl.ticker.MaxNLocator(
+                        nbins=num,
+                        min_n_ticks=num-1,
+                        steps=[1, 2, 2.5, 3, 5, 6, 8, 10]
+                    )
+
+                label_values = loc.tick_values(label_values_min,
+                                               label_values_max)
+                cond = ((label_values >= label_values_min) &
+                        (label_values <= label_values_max))
+                label_values = label_values[cond]
+                yarr = np.linspace(arr.min(), arr.max(), 256)
+                xarr = func(yarr)
+                ix = np.argsort(xarr)
+                values = np.interp(label_values, xarr[ix], yarr[ix])
+        elif num is not None and not label_values_are_numeric:
+            # Labels are not numerical so instead of interpolating
+            # just choose evenly distributed indexes instead:
+            if isinstance(num, mpl.ticker.Locator) or np.iterable(num):
+                raise ValueError(
+                    "`num` only supports integers for non-numeric labels."
+                )
+            cond = np.round(np.linspace(0, len(label_values) - 1, num))
+            cond = cond.astype(int)
+            values = values[cond]
             label_values = label_values[cond]
-            yarr = np.linspace(arr.min(), arr.max(), 256)
-            xarr = func(yarr)
-            ix = np.argsort(xarr)
-            values = np.interp(label_values, xarr[ix], yarr[ix])
 
+        # Some formatters requires set_locs:
+        if hasattr(fmt, "set_locs"):
+            fmt.set_locs(label_values)
+
+        # Default settings for handles, add or override with kwargs:
         kw = dict(markeredgewidth=self.get_linewidths()[0],
                   alpha=self.get_alpha())
         kw.update(kwargs)
 
         for val, lab in zip(values, label_values):
-            if prop == "colors":
-                color = self.cmap(self.norm(val))
-            elif prop == "sizes":
-                size = np.sqrt(val)
-                if np.isclose(size, 0.0):
-                    continue
```

Comment:
I've removed this becasue this wasn't being tested and made it really difficult to figure out the correct length without copying the code entirely. 

Even though adding 0 sized markers doesn't make sense to me the legend should show it if the user insists on it. 
I think instead of cleaning up in the legend `plt.scatter([0, 1]. [0, 1], s=[0, 1])` should've resulted in an error or warning instead.

Reply:
I think this is an API change though if I understand correctly?  People programatically set the size of markers all the time, and s=0 is not an unreasonable thing to do to mask out some markers that fall bellow a certain size.

Reply:
Sorry I guess the only change is possible here?  Not claiming I understand what this is doing, but it seems risky to just remove it....

Reply:
This check removed legend labels that have sizes close to 0 in the legend. So now that I removed it, everything that's in the plot will always be shown in the legend labels as well which makes sense to me.

`s=0` values are still in the plot being calculated with and slowing the down the plot. So using `s=0` to remove "bad" points from plot and legend seems like a bad habit that doesn't need encouragement. If you want to filter it should've been done before creating the scatter like:
```python
x = x[np.isclose(s, 0)]
y = y[np.isclose(s, 0)]
s = s[np.isclose(s, 0)]
ax.scatter(x, y, size=s)
```
 
One could also argue that we are just fixing a bug because: 
* It's not documented in the docs.
* It's not tested for. 
* Users can't change the `np.isclose()` parameters, so maybe there's some user out there working with small numbers and needs to change the parameters. So it's better to just let the user filter out points close to 0 before plotting the data.

We can undo this pretty easily. I will just find a seed that doesn't show this problem, and let it continue be hidden in the tests. But I of course am not a fan of that route.

Reply:
If it doesn't interfere w/ the rest of the PR I'd remove it and open a different PR, with a new API note, and tests appropriate for that case.  It seems an orthogonal discussion to be having!

Reply:
Done.